lib/AST/ASTPrinter.cpp

```diff
@@ -5156,6 +5203,103 @@ void PrintAST::visitErasureExpr(ErasureExpr *expr) {
 }
 
 void PrintAST::visitKeyPathExpr(KeyPathExpr *expr) {
+  if (expr->isObjC()) {
+    Printer << "#keyPath(";
+  }
+  
+  if (!expr->isObjC()) {
+    if (auto rootType = expr->getRootType()) {
+      Printer << "\\";
+      printTypeLoc(rootType);
+    } else {
+      visit(expr->getParsedRoot());
+    }
+  }
+  
+  auto components = expr->getComponents();
+  if (!components.empty()) {
+    for (auto& component: components) {
```

Comment:
```suggestion
    for (auto &component: components) {
```

Reply:
Also I think it would be great if you could factor out the component printing logic, which should make this function a bit easier to read, as e.g for the Obj-C case you could do:

```cxx
  if (expr->isObjC()) {
    Printer << "#keyPath(";
    printKeyPathComponents(expr->getComponents());
    Printer << ")";
    return;
  }
```

then you don't have to worry about intermixing it with the regular KeyPath printing logic.